### PR TITLE
Add a foreign-key constraint to the active_storage_attachments table for blobs

### LINKED ISSUE #32584: ActiveStorage foreign key enforcement and purge delete order
This issue is twofold:
1. The attachments table doesn't create a foreign key association in the database by default
2. 1 can't be done because of the order things happen in `#purge`

### Problem information
So, to ensure maximum enforcement in the database, I would want to the attachments table to have an foreign key to the blobs table. I can easily do this by specifying it in the migration:
````ruby
t.references :blob,     foreign_key: { to_table: :active_storage_blobs }, null: false
````

This works fine when initially adding an attachment.
But when I would delete it, the blob gets deleted first, then the attachment, creating a temporary case in which the FK wouldn't be used
````ruby
> some_model.image.purge
# ActiveStorage deletes the file, then immediately afterwards...
  ActiveStorage::Blob Destroy (1.3ms)  DELETE FROM "active_storage_blobs" WHERE "active_storage_blobs"."id" = $1  [["id", 1]]                                                                                                               
   (0.2ms)  ROLLBACK
ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  update or delete on table "active_storage_blobs" violates foreign key constraint "fk_rails_c3b3935057" on table "active_storage_attachments"                            
DETAIL:  Key (id)=(1) is still referenced from table "active_storage_attachments".
: DELETE FROM "active_storage_blobs" WHERE "active_storage_blobs"."id" = $1 
````
### Expected behavior
The purge process should complete successfully. However, I see that the process should go a bit differently than it currently does:
1. The attachment gets deleted. At this point, there is no association to whatever model you were originally looking at.
2. ActiveStorage deletes the actual file from storage.
3. After confirming the file's deletion, the blob gets deleted.

If it follows these steps, then the foreign key association is ok. Having the FK be a part of the default migration is another story, which I'm strongly suggesting with this.

### Actual behavior
Blob gets deleted first, then attachment. Foreign key is not possible with this arrangement, and if something goes wrong this can lead to database orphans. Also it appears to do both deletes in separate transactions in a situation where this is successful.

### System configuration
**Rails version**:
5.2.0
**Ruby version**:
2.4.4

Further information can be provided if necessary.

### PR DESCRIPTION
**Pull request by georgeclaghorn**
Enforce that purging a blob may not leave attachments behind. Closes #32584.

Note that this PR only adds the constraint in new apps. We’ll leave it up to maintainers of existing apps to add it if they wish.

### CODE REVIEW

File: activestorage/db/migrate/20170806125915_create_active_storage_tables.rb [hunk 1]
```diff
@@ -20,6 +20,7 @@ def change
       t.datetime :created_at, null: false
 
       t.index [ :record_type, :record_id, :name, :blob_id ], name: "index_active_storage_attachments_uniqueness", unique: true
+      t.foreign_key :active_storage_blobs, column: :blob_id
```
Comment (bogdanvlviv) [inline activestorage/db/migrate/20170806125915_create_active_storage_tables.rb]: @georgeclaghorn I'm wondering how users that migrate from `5.2` to `6.0` will know about this change. Maybe we should create a new migration for that or/and describe what users should do in order to get this change in http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-5-2-to-rails-6-0

Comment (kaspth) [reply activestorage/db/migrate/20170806125915_create_active_storage_tables.rb]: 👍, or we can perhaps copy an upgrade migration on `rails app:update`.

Comment (georgeclaghorn) [reply activestorage/db/migrate/20170806125915_create_active_storage_tables.rb]: On second thought, adding a migration on update would be nice.

**Comment by BKSpurgeon** [comment]
hi all thank you for your PR and contribution. was facing the same problem. my question: which version of rails will the above PR be incorporated in? chrs

**Comment by georgeclaghorn** [comment]
This’ll go into Rails 6.

**Comment by chaadow** [comment]
@geoffharcourt Hi could you add the `activestorage` label to this PR. As well as add the related issue to the 5.2.2 milestone

Just adding this constraint fixed the global issue when you update a record with failed validations.

- It ends up with an attachment pointing with a blob_id that does not exist anymore (because of the automatic purge_job), 
-Hence making helpers like `url_for` fail with a `undefined method #signed_id for nil:NilClass` 

PS : When you try to migrate your database with invalid data (for existing apps), the migration will fail. It would tell you that the `blob_id` with `(id)=(x)` column on `active_storage_attachments` table does not exist on the `active_storage_blobs` table.
This means that the blob was destroyed, but the attachments referencing the particular blob were not.
The solution i found is simple, in rails console you run : 
```ruby
ActiveStorage::Attachment.where(blob_id: x).destroy_all
```
Then re run the migration again, then in my case it would tell me about another attachment pointing to another non existent blob. 
So you repeat the operation (replacing `x` with the ID specified in the failed migration) until there are no more invalid data, and the migration passes. 

==> And for the time being, this makes activestorage work "normally" until `5.2.2` release or `6.0.0` which will remove the automatic upload on assignment, and the asynchrnous blob purging resulting in invalid data.

**Comment by geoffharcourt** [comment]
Hi @chaadow I think you meant to tag georgeclaghorn instead?

**Comment by chaadow** [comment]
@geoffharcourt Yes my bad!

**Comment by chaadow** [comment]
@georgeclaghorn My message was meant to be read by you 😅

**Comment by arthurnn** [comment]
thanks for this change. I wonder if there is an option, on Rails 7, to no add the FK to the migration?

**Pull request merged**